### PR TITLE
Refactor frexp API

### PR DESCRIPTION
**Pull request by steff456**
## Description

> What is the purpose of this pull request?

This pull request:

-   Refactors `frexp` package to support `assign` when the result needs to be overwrite a given array

## Related Issues

> Does this pull request have any related issues?

This PR is part of https://github.com/stdlib-js/todo/issues/1435

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/math/base/special/frexp/benchmark/benchmark.js [hunk 1]
```diff
@@ -50,7 +50,7 @@ bench( pkg, function benchmark( b ) {
 	b.end();
 });
 
-bench( pkg+'::memory_reuse', function benchmark( b ) {
+bench( pkg+'::assign', function benchmark( b ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/frexp/benchmark/benchmark.js]:
```suggestion
bench( pkg+':assign', function benchmark( b ) {
```

File: lib/node_modules/@stdlib/math/base/special/frexp/docs/repl.txt [hunk 2]
```diff
@@ -44,9 +41,47 @@
...
+        Output array index offset.
+
+    Returns
+    -------
+    out: Array|TypedArray|Object
+        A normalized fraction and an exponent.
+
+    Examples
+    --------
     > out = new {{alias:@stdlib/array/float64}}( 2 );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/frexp/docs/repl.txt]:
```suggestion
    > var out = new {{alias:@stdlib/array/float64}}( 2 );
```

New example block, so need to freshly declare any variables.

File: lib/node_modules/@stdlib/math/base/special/frexp/docs/repl.txt [hunk 3]
```diff
@@ -44,9 +41,47 @@
     > out = {{alias}}( {{alias:@stdlib/constants/float64/ninf}} )
     [ -Infinity, 0 ]
 
-    // Provide an output array:
+
+{{alias}}.assign( x, out, stride, offset )
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/frexp/docs/repl.txt:45]: L26 should be:

```
out: Array<number>
```

No longer can the output array be an Object or typed array.

File: lib/node_modules/@stdlib/math/base/special/frexp/docs/types/index.d.ts [hunk 4]
```diff
@@ -23,30 +23,75 @@
...
+	*
+	* @example
+	* var out = frexp( Infinity );
+	* // returns [ Infinity , 0 ]
+	*
+	* @example
+	* var out = frexp( -Infinity );
+	* // returns [ -Infinity , 0 ]
+	*/
+	( x: number ): Collection;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/frexp/docs/types/index.d.ts]:
```suggestion
	( x: number ): Array<number>;
```

For the main export, only an `Array` containing numbers is returned.

File: lib/node_modules/@stdlib/math/base/special/frexp/docs/types/test.ts [hunk 5]
```diff
@@ -19,23 +19,13 @@
...
-
-// The compiler throws an error if the function is provided an output array which is not a collection...
-{
-	frexp( 2, 4.0 ); // $ExpectError
-	frexp( false, 4.0 ); // $ExpectError
-	frexp( true, 4.0 ); // $ExpectError
-	frexp( {}, 4.0 ); // $ExpectError
 }
 
 // The compiler throws an error if the function is provided a last argument other than a number...
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/frexp/docs/types/test.ts]:
```suggestion
// The compiler throws an error if the function is provided an argument which is not a number...
```

File: lib/node_modules/@stdlib/math/base/special/frexp/docs/types/test.ts [hunk 6]
```diff
@@ -48,19 +38,86 @@ import { Collection } from '@stdlib/types/object';
...
+
+// The compiler throws an error if the function is provided an output array that is not an array-like object of numbers...
+{
+	frexp( '5', 3.14e-319 ); // $ExpectError
+	frexp( true, 3.14e-319 ); // $ExpectError
+	frexp( false, 3.14e-319 ); // $ExpectError
+	frexp( null, 3.14e-319 ); // $ExpectError
+	frexp( {}, 3.14e-319 ); // $ExpectError
+	frexp( ( x: number ): number => x, 3.14e-319 ); // $ExpectError
+}
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/frexp/docs/types/test.ts]:
```suggestion
```

These tests are no longer needed.

File: lib/node_modules/@stdlib/math/base/special/frexp/lib/main.js [hunk 7]
```diff
@@ -67,11 +55,8 @@ var fcn = require( './frexp.js' );
 * var out = frexp( -Infinity );
 * // returns [ -Infinity , 0 ]
 */
-function frexp( out, x ) {
-	if ( arguments.length === 1 ) {
-		return fcn( [ 0.0, 0 ], out );
-	}
-	return fcn( out, x );
+function frexp( x ) {
+	return fcn( x, [0.0, 0.0], 1, 0 );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/frexp/lib/main.js]:
```suggestion
	return fcn( x, [ 0.0, 0 ], 1, 0 );
```

File: lib/node_modules/@stdlib/math/base/special/frexp/lib/main.js [hunk 8]
```diff
@@ -20,15 +20,14 @@
...
 
 
 // MAIN //
 
 /**
 * Splits a double-precision floating-point number into a normalized fraction and an integer power of two.
 *
-* @param {(Array|TypedArray|Object)} [out] - output array
 * @param {number} x - input value
 * @returns {(Array|TypedArray|Object)} output array
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/frexp/lib/main.js]:
```suggestion
* @returns {Array<number>} output array
```

File: lib/node_modules/@stdlib/math/base/special/frexp/test/test.assign.js [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+
+	out = new Float64Array( 2 );
+	f = frexp( NaN, out, 1, 0 );
+	t.equal( f, out, 'returns output array' );
+	t.equal( isnan( f[0] ), true, 'first element is NaN' );
+	t.equal( f[ 1 ], 0, 'second element is 0' );
+	t.end();
+});
+
+tape( 'the function supports provided an output array', function test( t ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/frexp/test/test.assign.js]:
```suggestion
tape( 'the function supports providing an output array', function test( t ) {
```

File: lib/node_modules/@stdlib/math/base/special/frexp/test/test.assign.js [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+	f = frexp( 4.0, out, 1, 0 );
+
+	t.equal( f, out, 'returns output array' );
+	t.equal( f[ 0 ], 0.5, 'has expected first element' );
+	t.equal( f[ 1 ], 3, 'has expected second element' );
+
+	t.end();
+});
+
+tape( 'the function supports provided an output typed array', function test( t ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/frexp/test/test.assign.js]:
```suggestion
tape( 'the function supports providing an output typed array', function test( t ) {
```

**Review by kgryte — CHANGES_REQUESTED** [review]

File: lib/node_modules/@stdlib/math/base/special/frexp/docs/repl.txt (hunk 3 above)
Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/special/frexp/docs/repl.txt:45]: @steff456 I think this is the main thing to be changed.

**Comment by kgryte** [comment]
@steff456 As part of this refactor, did you identify any dependents (i.e., packages using `frexp`) that would need to be updated as a result of refactoring to an `assign` API?

File: lib/node_modules/@stdlib/math/base/special/frexp/lib/assign.js [hunk 11]
```diff
@@ -46,58 +46,60 @@ var WORDS = [ 0, 0 ]; // WARNING: not thread safe
...
 	) {
-		out[ 0 ] = x;
-		out[ 1 ] = 0;
+		out[ offset ] = x;
+		out[ offset + stride ] = 0;
 		return out;
 	}
 	// If `x` is subnormal, normalize it...
-	normalize( x, X, 1, 0 );
+	normalize( x, X, stride, offset );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/frexp/lib/assign.js]:
```suggestion
	normalize( x, X, 1, 0 );
```

The stride and offset should remain `1` and `0`, as these arguments refer to `X`, the workspace, not `out`.

File: lib/node_modules/@stdlib/math/base/special/frexp/lib/assign.js [hunk 12]
```diff
@@ -46,58 +46,60 @@ var WORDS = [ 0, 0 ]; // WARNING: not thread safe
...
+		out[ offset ] = x;
+		out[ offset + stride ] = 0;
 		return out;
 	}
 	// If `x` is subnormal, normalize it...
 	normalize( x, X, 1, 0 );
 
 	// Extract the exponent from `x` and add the normalization exponent:
-	exp = floatExp( X[0] ) + X[ 1 ] + 1;
+	exp = floatExp( X[ offset ] ) + X[ offset + stride ] + 1;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/frexp/lib/assign.js]:
```suggestion
	exp = floatExp( X[ 0 ] ) + X[ 1 ] + 1;
```

The workspace is distinct from the output array.

File: lib/node_modules/@stdlib/math/base/special/frexp/lib/assign.js [hunk 13]
```diff
@@ -46,58 +46,60 @@ var WORDS = [ 0, 0 ]; // WARNING: not thread safe
...
 
 	// Break `x` into two unsigned 32-bit integers (higher and lower order words):
-	toWords( WORDS, X[ 0 ] );
-	high = WORDS[ 0 ];
+	toWords( WORDS, X[ offset ] );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/frexp/lib/assign.js]:
```suggestion
	toWords( WORDS, X[ 0 ] );
```

File: lib/node_modules/@stdlib/math/base/special/frexp/lib/assign.js [hunk 14]
```diff
@@ -46,58 +46,60 @@ var WORDS = [ 0, 0 ]; // WARNING: not thread safe
...
+	high = WORDS[ offset + stride ];
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/frexp/lib/assign.js]:
```suggestion
	high = WORDS[ 1 ];
```

The workspace is distinct from the output array.

File: lib/node_modules/@stdlib/math/base/special/frexp/lib/assign.js [hunk 15]
```diff
@@ -106,10 +108,10 @@ function frexp( out, x ) {
 	high |= SET_EXP_MASK;
 
 	// Create a new floating-point number:
-	x = fromWords( high, WORDS[ 1 ] );
+	x = fromWords( high, WORDS[ offset + stride ] );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/frexp/lib/assign.js]:
```suggestion
	x = fromWords( high, WORDS[ 1 ] );
```

Workspace is distinct from the output array.

File: lib/node_modules/@stdlib/math/base/special/frexp/lib/assign.js [hunk 16]
```diff
@@ -46,58 +46,60 @@ var WORDS = [ 0, 0 ]; // WARNING: not thread safe
...
 	normalize( x, X, 1, 0 );
 
 	// Extract the exponent from `x` and add the normalization exponent:
-	exp = floatExp( X[0] ) + X[ 1 ] + 1;
+	exp = floatExp( X[ 0 ] ) + X[ 1 ] + 1;
 
 	// Break `x` into two unsigned 32-bit integers (higher and lower order words):
 	toWords( WORDS, X[ 0 ] );
-	high = WORDS[ 0 ];
+	high = WORDS[ 1 ];
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/frexp/lib/assign.js]:
```suggestion
	high = WORDS[ 0 ];
```

File: lib/node_modules/@stdlib/math/base/special/frexp/lib/assign.js [hunk 17]
```diff
@@ -46,54 +46,56 @@ var WORDS = [ 0, 0 ]; // WARNING: not thread safe
...
+		out[ offset ] = x;
+		out[ offset + stride ] = 0;
 		return out;
 	}
 	// If `x` is subnormal, normalize it...
 	normalize( x, X, 1, 0 );
 
 	// Extract the exponent from `x` and add the normalization exponent:
-	exp = floatExp( X[0] ) + X[ 1 ] + 1;
+	exp = floatExp( X[ 0 ] ) + X[ 1 ] + 1;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/frexp/lib/assign.js]:
```suggestion
	exp = floatExp( X[0] ) + X[ 1 ] + 1;
```

File: lib/node_modules/@stdlib/math/base/special/frexp/test/test.assign.js [hunk 18]
```diff
@@ -0,0 +1,315 @@
...
+		if ( randu() < 0.5 ) {
+			sign = -1.0;
+		} else {
+			sign = 1.0;
+		}
+		frac = randu() * 10.0;
+		exp = round( randu()*614.0 ) - 307;
+		x = sign * frac * pow( 10.0, exp );
+		out = new Float64Array( 2 );
+		f = frexp( x[i], out, 1, 0 );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/frexp/test/test.assign.js]:
```suggestion
		f = frexp( x, out, 1, 0 );
```

File: lib/node_modules/@stdlib/math/base/special/frexp/test/test.assign.js [hunk 19]
```diff
@@ -0,0 +1,315 @@
...
+		if ( randu() < 0.5 ) {
+			sign = -1.0;
+		} else {
+			sign = 1.0;
+		}
+		frac = randu() * 10.0;
+		exp = round( randu()*616.0 ) - 308;
+		x = sign * frac * pow( 10.0, exp );
+		out = new Float64Array( 2 );
+		f = frexp( x[i], out, 1, 0 );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/frexp/test/test.assign.js]:
```suggestion
		f = frexp( x, out, 1, 0 );
```

File: lib/node_modules/@stdlib/math/base/special/frexp/test/test.assign.js [hunk 20]
```diff
@@ -0,0 +1,315 @@
...
+		frac = randu() * 10.0;
+		exp = round( randu()*614.0 ) - 307;
+		x = sign * frac * pow( 10.0, exp );
+		out = new Float64Array( 2 );
+		f = frexp( x, out, 1, 0 );
+
+		// Compute the absolute value of the normalized fraction:
+		f = abs( f[ 0 ] );
+
+		t.equal( f, out, 'returns output array' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/frexp/test/test.assign.js]:
```suggestion
		t.equal( f, out, 'returns output array' );

		// Compute the absolute value of the normalized fraction:
		f = abs( f[ 0 ] );
```

File: lib/node_modules/@stdlib/math/base/special/frexp/test/test.assign.js [hunk 21]
```diff
@@ -0,0 +1,314 @@
...
+		out = new Float64Array( 2 );
+		f = frexp( x, out, 1, 0 );
+
+		if ( f[ 1 ] > BIAS ) {
+			f = f[ 0 ] * pow( 2.0, BIAS ) * pow( 2.0, f[1]-BIAS );
+		} else {
+			f = f[ 0 ] * pow( 2.0, f[ 1 ] );
+		}
+		t.equal( f, out, 'returns output array' );
+		t.equal( f, x, 'frac * 2^exp equals ' + x );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/frexp/test/test.assign.js]:
```suggestion
		t.equal( f, out, 'returns output array' );

		if ( f[ 1 ] > BIAS ) {
			f = f[ 0 ] * pow( 2.0, BIAS ) * pow( 2.0, f[1]-BIAS );
		} else {
			f = f[ 0 ] * pow( 2.0, f[ 1 ] );
		}
		t.equal( f, x, 'frac * 2^exp equals ' + x );
```

**Comment by steff456** [comment]
> @steff456 As part of this refactor, did you identify any dependents (i.e., packages using frexp) that would need to be updated as a result of refactoring to an assign API?

I need to double check!

**Comment by steff456** [comment]
Update: there's no dependents that need to be updated 😉

**Comment by kgryte** [comment]
@steff456 Thanks for checking!

**Comment by kgryte** [comment]
@steff456 Aren't `@stdlib/stats/incr/prod` and `@stdlib/stats/incr/mprod` affected by the changes introduced in this PR?

**Comment by steff456** [comment]
You are right! I'm not sure why my editor was excluding files in the search, but now it is fixed! Thanks for double checking on your side @kgryte :)

**Comment by Planeshifter** [comment]
@steff456 I think any `node_modules` folders are being ignored by default in most code editors. I remember having had to configure my workspace to make sure that the `lib/node_modules/@stdlib` folder and its sub-folders are not ignored.

File: lib/node_modules/@stdlib/stats/incr/mprod/lib/main.js [hunk 22]
```diff
@@ -42,7 +42,7 @@ var format = require( '@stdlib/string/format' );
 */
 function product( workspace, acc, x ) {
 	// Split the incoming value into a normalized fraction and exponent:
-	frexp( workspace, x );
+	frexp.assign( x, workspace );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/stats/incr/mprod/lib/main.js]:
```suggestion
	frexp.assign( x, workspace, 1, 0 );
```

Missing stride and offset.

File: lib/node_modules/@stdlib/stats/incr/mprod/lib/main.js [hunk 23]
```diff
@@ -52,7 +52,7 @@ function product( workspace, acc, x ) {
 
 	// Ensure fraction remains normalized to avoid overflow/underflow...
 	if ( acc.frac > -0.5 && acc.frac < 0.5 ) {
-		frexp( workspace, acc.frac );
+		frexp( acc.frac, workspace );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/stats/incr/mprod/lib/main.js]:
```suggestion
		frexp( acc.frac, workspace, 1, 0 );
```

Missing stride and offset.

File: lib/node_modules/@stdlib/stats/incr/prod/lib/main.js [hunk 24]
```diff
@@ -86,7 +86,7 @@ function incrprod() {
 			return ( prod === void 0 ) ? null : prod;
 		}
 		// Split the incoming value into a normalized fraction and exponent:
-		frexp( PARTS, x );
+		frexp.assign( x, PARTS );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/stats/incr/prod/lib/main.js]:
```suggestion
		frexp.assign( x, PARTS, 1, 0 );
```

Missing stride and offset.

File: lib/node_modules/@stdlib/stats/incr/prod/lib/main.js [hunk 25]
```diff
@@ -96,7 +96,7 @@ function incrprod() {
 
 		// Ensure fraction remains normalized to avoid overflow/underflow...
 		if ( frac > -0.5 && frac < 0.5 ) {
-			frexp( PARTS, frac );
+			frexp.assign( frac, PARTS );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/stats/incr/prod/lib/main.js]:
```suggestion
			frexp.assign( frac, PARTS, 1, 0 );
```

Missing stride and offset.

**Review by kgryte — APPROVED** [review]
LGTM. Thanks, @steff456!

**Review by Planeshifter — APPROVED** [review]

**Pull request merged**